src/lib/mdns/Resolver_ImplMinimalMdns.cpp

```diff
@@ -491,47 +508,86 @@ CHIP_ERROR MinMdnsResolver::BrowseNodes(DiscoveryType type, DiscoveryFilter filt
 
 CHIP_ERROR MinMdnsResolver::ResolveNodeId(const PeerId & peerId, Inet::IPAddressType type)
 {
-    mDiscoveryType                    = DiscoveryType::kOperational;
-    System::PacketBufferHandle buffer = System::PacketBufferHandle::New(kMdnsMaxPacketSize);
-    ReturnErrorCodeIf(buffer.IsNull(), CHIP_ERROR_NO_MEMORY);
+    mDiscoveryType = DiscoveryType::kOperational;
+    mActiveResolves.MarkPending(peerId);
 
-    QueryBuilder builder(std::move(buffer));
-    builder.Header().SetMessageId(0);
+    return SendPendingResolveQueries();
+}
+
+CHIP_ERROR MinMdnsResolver::ScheduleResolveRetries()
+{
+    ReturnErrorCodeIf(mSystemLayer == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    mSystemLayer->CancelTimer(&ResolveRetryCallback, this);
 
+    Optional<uint32_t> delayMs = mActiveResolves.GetMsUntilNextExpectedResponse();
+
+    if (!delayMs.HasValue())
     {
-        char nameBuffer[64] = "";
-
-        // Node and fabricid are encoded in server names.
-        ReturnErrorOnFailure(MakeInstanceName(nameBuffer, sizeof(nameBuffer), peerId));
-
-        const char * instanceQName[] = { nameBuffer, kOperationalServiceName, kOperationalProtocol, kLocalDomain };
-        Query query(instanceQName);
-
-        query
-            .SetClass(QClass::IN)       //
-            .SetType(QType::ANY)        //
-            .SetAnswerViaUnicast(false) //
-            ;
-
-        // NOTE: type above is NOT A or AAAA because the name searched for is
-        // a SRV record. The layout is:
-        //    SRV -> hostname
-        //    Hostname -> A
-        //    Hostname -> AAAA
-        //
-        // Query is sent for ANY and expectation is to receive A/AAAA records
-        // in the additional section of the reply.
-        //
-        // Sending a A/AAAA query will return no results
-        // Sending a SRV query will return the srv only and an additional query
-        // would be needed to resolve the host name to an IP address
-
-        builder.AddQuery(query);
+        return CHIP_NO_ERROR;
     }
 
-    ReturnErrorCodeIf(!builder.Ok(), CHIP_ERROR_INTERNAL);
+    return mSystemLayer->StartTimer(delayMs.Value(), &ResolveRetryCallback, this);
+}
 
-    return GlobalMinimalMdnsServer::Server().BroadcastSend(builder.ReleasePacket(), kMdnsPort);
+void MinMdnsResolver::ResolveRetryCallback(System::Layer *, void * self)
+{
+    reinterpret_cast<MinMdnsResolver *>(self)->SendPendingResolveQueries();
+}
+
+CHIP_ERROR MinMdnsResolver::SendPendingResolveQueries()
+{
+    while (true)
+    {
+        Optional<PeerId> peerId = mActiveResolves.NextScheduledPeer();
+
+        if (!peerId.HasValue())
+        {
+            break;
+        }
+
+        System::PacketBufferHandle buffer = System::PacketBufferHandle::New(kMdnsMaxPacketSize);
+        ReturnErrorCodeIf(buffer.IsNull(), CHIP_ERROR_NO_MEMORY);
+
+        QueryBuilder builder(std::move(buffer));
+        builder.Header().SetMessageId(0);
+
+        {
+            char nameBuffer[64] = "";
```

Comment:
Do we have a define somewhere that tells us how long this buffer needs to be?

Reply:
This was existing code that I moved around, will have to search for a define.